Core tracedb functionality.

debris opened the pull request:
separated core tracedb functionality from pr #924.

99% of this pr is a new code. Removed lines are mostly variables/functions renaming. Most of the new code is added to `ethcore/trace` module. Changes in this pr won't store traces in the database, but they provide everything that is required to do so.

#### this pr includes:

- fixed encoding of `0u8` (#992 ,merged)
- import route contains ommited blocks (#994 ,merged)
- `trace::ExecutiveTracer` and `trace::NoopTracer` moved to separate files
- traces rlp encoding / decoding
- traces bloom
- traces filter
- tracedb
- tracedb config
- removed `have_tracing` from blockchain
- blockchain bloom filtering is done using external crate [bloomchain](https://github.com/debris/bloomchain)
- turned on tracing of transactions to builtins. It simplifies querying traces, cause number of traces is always equal to number of transactions in given block.

#### this pr does not include:

- client api for tracing
- actual storing traces in database
- tracing cli
- traces rpc

tomusdrw commented on ethcore/src/trace/db.rs:
```diff
@@ -0,0 +1,573 @@
+	pub fn new(mut config: Config, path: &Path, extras: Arc<T>) -> Self {
+		let mut tracedb_path = path.to_path_buf();
+		tracedb_path.push("tracedb");
+		let tracesdb = Database::open_default(tracedb_path.to_str().unwrap()).unwrap();
```
Will path be automatically created? (`path` and `tracedb_path`)

debris replied:
yes

arkpar replied:
what about versioning?

debris replied:
added here: https://github.com/ethcore/parity/pull/996/commits/0b2855d58a82f3b773aec17571ecb52ae9a9a0a1

gavofyork replied:
what's the idea behind having a separate DB? are we going to start having loads of different DBs now? VM tracing, and fat db (i.e. hash preimages) are both independent additions. is there a storage/memory/compute/file-handle cost to loads of open databases?

tomusdrw commented on ethcore/src/trace/db.rs:
```diff
@@ -0,0 +1,573 @@
+		};
+
+		// compare it with the current option.
+		let tracing = match (tracing_was_enabled, config.enabled) {
```
Why exactly do we need `Option` in both `tracing_was_enabled`, `config.enabled`? Are `None`s representing defaults?

debris replied:
Yes, they represent defaults. Do you think there should be a another type here?

tomusdrw commented on ethcore/src/trace/db.rs:
```diff
@@ -0,0 +1,573 @@
+		tx_number: usize
+	) -> Vec<LocalizedTrace> {
+		let tx_hash = self.extras.transaction_hash(block_number, tx_number)
+		.expect("Expected to find transaction hash. Database is probably malformed");
```
Indent?

tomusdrw commented on ethcore/src/trace/db.rs:
```diff
@@ -0,0 +1,573 @@
+		.expect("Expected to find transaction hash. Database is probably malformed");
+
+		let flat_traces: Vec<FlatTrace> = traces.into();
+		flat_traces.into_iter()
```
Implementing `traces.into()` directly to `Iterator` would simplify couple of `into().into_iter()` chains.

debris replied:
I can't do that, cause flat trace does not have all required data - `trace_number`, `transaction_number`, `transaction_hash`, `block_number` and `block_hash`.

tomusdrw commented on ethcore/src/trace/db.rs:
```diff
@@ -0,0 +1,573 @@
+
+impl<T> TraceDatabase for Tracedb<T> where T: DatabaseExtras {
+	fn tracing_enabled(&self) -> bool {
+		self.config.enabled.expect("Auto tracing hasn't been properly configured.")
```
Why do we need `config.enabled` to be `Option` again?

debris replied:
What do you mean exactly?

The only way to avoid `unwrapping` option here is to store `enabled` field as a separate property. Do you think it is better?

tomusdrw commented on ethcore/src/trace/db.rs:
```diff
@@ -0,0 +1,573 @@
+		self.config.enabled.expect("Auto tracing hasn't been properly configured.")
+	}
+
+	/// Traces of impor request's enacted blocks are expected to be already in database
```
tiny typo

tomusdrw commented on ethcore/src/trace/db.rs:
```diff
@@ -0,0 +1,573 @@
+		self.extras.block_hash(block_number)
+			.and_then(|block_hash| self.traces(&block_hash)
+				.map(FlatBlockTraces::from)
+				.map(Into::<Vec<FlatTransactionTraces>>::into)
```
3 lines above are quite common (6 occurrences) (`traces.map(FlatBlock).map(Into::Vec<...`) maybe it should be separated into method?

tomusdrw commented on ethcore/src/trace/db.rs:
```diff
@@ -0,0 +1,573 @@
+
+	#[test]
+	#[should_panic]
+	fn test_invalid_reopeining_db() {
```
typo

tomusdrw commented on ethcore/src/trace/filter.rs:
```diff
@@ -0,0 +1,255 @@
+	/// Block range.
+	pub range: Range<usize>,
+
+	/// From address. If empty, match all, if not, match one of the values.
```
Why not encoding information from docs into enum type?

debris replied:
If I use enum type eg.

```rust
pub enum FilterAddresses {
  All
  Some(HashSet<Address>)
}
```

I would have to check if `Some` is empty... But having `pub struct FilterAddresses(HashSet<Address>)` sounds good :)

tomusdrw commented on ethcore/src/trace/filter.rs:
```diff
@@ -0,0 +1,255 @@
+	pub range: Range<usize>,
+
+	/// From address. If empty, match all, if not, match one of the values.
+	pub from_address: Vec<Address>,
```
You mostly check `.contains` and don't care about order - using `HashSet` would be more appropriate here.

tomusdrw commented on ethcore/src/trace/noop_tracer.rs:
```diff
@@ -0,0 +1,66 @@
+	}
+
+	fn trace_call(&mut self, call: Option<Call>, _: U256, output: Option<Bytes>, _: usize, _: Vec<Trace>,
+				  _: bool) {
```
Same line?

tomusdrw commented:
Had a quick look, but I need more time to dig into the logic. Would be cool for some more experienced guys to look into general design / logic.

arkpar commented on ethcore/src/executive.rs:
```diff
@@ -245,30 +246,51 @@ impl<'a> Executive<'a> {
 		}
 		trace!("Executive::call(params={:?}) self.env_info={:?}", params, self.info);
 
+		let delegate_call = params.code_address != params.address;
```
Is this correct? What if contract `A` calls `B` which calls back `A`?

debris replied:
Ahh... you are right. But this edge case is not trivial to fix. With our current code design `Executive` does not know if `A` callcode `B` which callcode `A`, because it's irrelevant to `A` state (the result of this operation is the same as `A` calling itself).

I suggest leaving this as it is and fixing later, cause it may require additional changes in evm (exposing extra data).

gavofyork replied:
delegate calls are not traced further - if A delegatecalls B delegatecalls A, then, correctly, nothing will be traced.

arkpar commented on ethcore/src/state.rs:
```diff
@@ -550,6 +552,10 @@ fn should_trace_basic_call_transaction() {
+// call tracing turned on to make
+// number of transactions and traces in a block equal
+// debris
 fn should_not_trace_call_transaction_to_builtin() {
```
Still not clear why the test is disabled

arkpar commented on ethcore/src/trace/bloom.rs:
```diff
@@ -0,0 +1,119 @@
+		use std::ptr;
+		let mut hash = H256::default();
+		unsafe {
+			ptr::copy(&[self.level, self.index] as *const usize as *const u8, hash.as_mut_ptr(), 16);
```
This won't work for 32-bit build. usize is pointer sized. 
Why not just use two 32-bit values and combine them into one 64-bit and return it?

arkpar commented on ethcore/src/trace/bloom.rs:
```diff
@@ -0,0 +1,119 @@
+
+impl TraceGroupPosition {
+	/// Returns trace group position represented as a 256 bit hash.
+	pub fn hash(&self) -> H256 {
```
Does it need to be H256?

arkpar commented on ethcore/src/trace/config.rs:
```diff
@@ -0,0 +1,40 @@
+pub struct Config {
+	/// Indicates if tracing should be enabled or not.
+	/// If it's None, it will be automatically configured.
+	pub enabled: Option<bool>,
```
Better just use a 3-value enum

arkpar commented on ethcore/src/trace/db.rs:
```diff
@@ -0,0 +1,575 @@
+}
+
+impl Key<BlockTracesBloomGroup> for TraceGroupPosition {
+	fn key(&self) -> H264 {
```
Key size should be variable. We don't want to bloat the database. Consider using ElasticArray

arkpar commented on ethcore/src/trace/db.rs:
```diff
@@ -0,0 +1,575 @@
+		tx_number: usize
+	) -> Vec<LocalizedTrace> {
+		let tx_hash = self.extras.transaction_hash(block_number, tx_number)
+			.expect("Expected to find transaction hash. Database is probably malformed");
```
more like "corrupted"

arkpar commented on ethcore/src/trace/executive_tracer.rs:
```diff
@@ -0,0 +1,108 @@
+		Some(vec![])
+	}
+
+	fn trace_call(&mut self, call: Option<Call>, gas_used: U256, output: Option<Bytes>, depth: usize, subs:
```
Why is call optional if the function expects it to be `Some` anyway?

debris replied:
I use options to avoid unnecessary memory copying of input / output. There is also `NoopTracer` which is used when tracing is disabled and it uses only `None` variant to enum.

arkpar commented on ethcore/src/trace/filter.rs:
```diff
@@ -0,0 +1,284 @@
+/// Addresses filter.
+///
+/// Used to create bloom possibilities and match filters.
+pub struct AddressesFilter(Vec<Address>);
```
consider using `HashSet`

debris replied:
we discussed it. it's not a performance issue here, vector is likely to have just a few elements most of the time (1-4).

arkpar commented on ethcore/src/trace/bloom.rs:
```diff
@@ -0,0 +1,106 @@
+#[derive(PartialEq, Eq, Hash, Clone, Debug)]
+pub struct TraceGroupPosition {
+	/// Bloom level.
+	pub level: u8,
```
Will it always be 256 levels max?

debris replied:
yes, currently we use just 3

NikVolf commented:
is it supposed to get into 1.1?

debris commented:
yes

gavofyork commented on ethcore/src/state.rs:
```diff
@@ -571,7 +573,21 @@ fn should_not_trace_call_transaction_to_builtin() {
 	let result = state.apply(&info, engine.deref(), &t, true).unwrap();
 
-	assert_eq!(result.trace, None);
+	assert_eq!(result.trace, Some(Trace {
```
what's going on here? expectation changed since previous implementation?

debris replied:
I turned on tracing of transactions to builtins. It simplifies querying traces by transaction hash, cause number of traces is always equal to number of transactions in given block.

gavofyork replied:
that's not a particularly great idea - i seem to remember christian putting in a lot of `CALL`s to the identity contract (0x4) in solidity, which uses it as a `memcpy`; all of these would get caught and will quite possibly severely bloat the DB.

gavofyork commented on ethcore/src/state.rs:
```diff
@@ -599,18 +615,32 @@ fn should_not_trace_subcall_transaction_to_builtin() {
 			output: vec![]
 		}),
-		subs: vec![]
+		subs: vec![Trace {
```
what's going on here? expectation changed since previous implementation?

gavofyork commented on ethcore/src/trace/config.rs:
```diff
@@ -0,0 +1,75 @@
+			(Switch::Off, Switch::Auto) => Ok(Switch::Off),
+			(Switch::Off, Switch::Off) => Ok(Switch::Off),
+			(Switch::Auto, Switch::On) => Ok(Switch::On),
+			_ => Ok(Switch::Off),
```
since this is default, why not remove the other arms that map to `Ok(Switch::Off)`?

gavofyork commented on ethcore/src/trace/config.rs:
```diff
@@ -0,0 +1,75 @@
+	pub fn as_bool(&self) -> bool {
+		match *self {
+			Switch::On => true,
+			Switch::Off | Switch::Auto => false,
```
`Switch::Auto` effectively mapped to `true` in `turn_to` - why is it `false` here?

gavofyork commented on ethcore/src/trace/db.rs:
```diff
@@ -0,0 +1,589 @@
+const TRACE_DB_VER: &'static [u8] = b"1.0";
+
+#[derive(Debug, Copy, Clone)]
+pub enum TracedbIndex {
```
DB is separate word generally. Alter name to `TraceDBIndex`.

gavofyork commented on ethcore/src/trace/db.rs:
```diff
@@ -0,0 +1,589 @@
+	}
+}
+
+/// Fat database.
```
`Fat DB` is generally reserved for storing the prehash images of the storage/state keys. Present functionality is "transaction tracing".

debris replied:
This is obsolete comment. will fix it

gavofyork commented on ethcore/src/trace/db.rs:
```diff
@@ -0,0 +1,589 @@
+	}
+
+	#[test]
+	fn test_reopeining_db_with_tracing_on() {
```
misspelling.

gavofyork commented on ethcore/src/trace/executive_tracer.rs:
```diff
@@ -0,0 +1,108 @@
+		let trace = Trace {
+			depth: depth,
+			subs: subs,
+			action: Action::Call(call.expect("Trace call expected to be Some.")),
```
unacceptable panic. either PROVE it cannot happen or handle it properly.

debris replied:
Options are used here just to avoid unnecessary memory allocations. `expect` indicates, that the `call` object was created incorrectly, probably with another Tracer and should fail verbosely - during tests.

*this code is already merged and part of the master branch. This code is marked as 'new' only because I moved the implementation to another file.

gavofyork replied:
then it should never have been merged. we have a strict policy on panickers.

PROVE or REMOVE.

gavofyork commented on ethcore/src/trace/executive_tracer.rs:
```diff
@@ -0,0 +1,108 @@
+			action: Action::Call(call.expect("Trace call expected to be Some.")),
+			result: Res::Call(CallResult {
+				gas_used: gas_used,
+				output: output.expect("Trace call output expected to be Some.")
```
unacceptable panic. either PROVE it can NEVER happen or handle it properly.

gavofyork commented on ethcore/src/trace/executive_tracer.rs:
```diff
@@ -0,0 +1,108 @@
+		let trace = Trace {
+			depth: depth,
+			subs: subs,
+			action: Action::Create(create.expect("Trace create expected to be Some.")),
```
unacceptable panic. either PROVE it can NEVER happen or handle it properly.

gavofyork commented on ethcore/src/trace/executive_tracer.rs:
```diff
@@ -0,0 +1,108 @@
+			action: Action::Create(create.expect("Trace create expected to be Some.")),
+			result: Res::Create(CreateResult {
+				gas_used: gas_used,
+				code: code.expect("Trace create code expected to be Some."),
```
unacceptable panic. either PROVE it can NEVER happen or handle it properly.

gavofyork commented on ethcore/src/trace/executive_tracer.rs:
```diff
@@ -0,0 +1,108 @@
+		let trace = Trace {
+			depth: depth,
+			subs: subs,
+			action: Action::Call(call.expect("Trace call expected to be Some.")),
```
unacceptable panic. either PROVE it can NEVER happen or handle it properly.

gavofyork commented on ethcore/src/trace/executive_tracer.rs:
```diff
@@ -0,0 +1,108 @@
+		let trace = Trace {
+			depth: depth,
+			subs: subs,
+			action: Action::Create(create.expect("Trace create expected to be Some.")),
```
unacceptable panic. either PROVE it can NEVER happen or handle it properly.

gavofyork commented on ethcore/src/trace/filter.rs:
```diff
@@ -0,0 +1,284 @@
+	use basic_types::LogBloom;
+
+	#[test]
+	fn empty_trace_filter_bloom_possibilies() {
```
spelling

gavofyork commented on ethcore/src/trace/flat.rs:
```diff
@@ -0,0 +1,183 @@
+			})
+			.collect::<Vec<_>>();
+
+		let ordered = FlatTrace {
```
probably want an address here.

i.e. if the trace is:
A
  CALLs B
    CALLs G
  CALLs C
    CALLs G
B
  CALLs G

then it should look something like:

[ {A, [0]}, {B, [0, 0]}, {G, [0, 0, 0]}, {C, [0, 1]}, {G, [0, 1, 0]}, {B, [1]}, {G, [1, 0]} ]

debris replied:
> probably want an "address" here, giving the exactly location in the call trace [index in root, index in first CALL, index in second CALL, ...] .

`FlatTrace` is not a part of public `trace` module api. In public api, there is another data structure called [`LocalizedTrace`](https://github.com/ethcore/parity/pull/997/files#diff-9e4ac8e6da1fc0d8eceebe2cb09f9fccR23). It has `trace_number`, `parent` (trace number of a parent) and `children` (trace numbers of children). Trace number is just a number of trace within given transaction. I think it should be sufficient.

> since depth is not enough to figure out sub-call order without an ordered depth-first enumeration of the entire trace, which may not be available (though i guess it could be done at time of retrieval rather than insertion).

Right now `flattening` is done at retrieval time.

gavofyork replied:
fair enough - but parent/children feels a bit clumsy as it requires unique indexing for the call tree's nodes. sticking to vector-addressing scheme you avoid such a need for indexing. can you think of a use case where one would be better than the other?

debris replied:
You are right, I will go with `vector-addressing scheme`, cause it provides more information, eg. all `parent` traces.

gavofyork commented on ethcore/src/trace/trace.rs:
```diff
@@ -64,9 +110,44 @@ impl From<ActionParams> for TraceCall {
+		s.append(&self.to);
+		s.append(&self.value);
+		s.append(&self.gas);
+		s.append(&self.input);
```
what happened to `output`?

debris commented:
changes in last few prs:

- fixed typos, misspellings
- trace calls to builtins only when depth == 0
- Renamed Tracedb to TraceDB, TracedbIndex to TraceDBIndex.
- Proves in tracers assertions.
- fixed switch `turn_to` logic.
- vector-addressing scheme for Traces

gavofyork commented on ethcore/src/trace/filter.rs:
```diff
@@ -271,6 +268,8 @@ mod tests {
 				input: vec![0x5],
 			}),
 			result: Res::FailedCall,
+			trace_address: vec![0],
```
is this correct? or should it be `vec![]`?

gavofyork commented on ethcore/src/trace/db.rs:
```diff
@@ -512,7 +498,8 @@ mod tests {
 			}),
 			result: Res::FailedCall,
-			trace_number: 0,
+			trace_address: vec![0],
```
is this correct? or should it be `vec![]`?

gavofyork commented on ethcore/src/trace/flat.rs:
```diff
@@ -169,15 +164,25 @@ mod tests {
-		assert_eq!(ordered_traces[4].children, vec![]);
+		//assert_eq!(ordered_traces[0].parent, None);
+		//assert_eq!(ordered_traces[0].children, vec![1, 4]);
+		assert_eq!(ordered_traces[0].trace_address, vec![0]);
```
should be `vec![]`, i think.

gavofyork commented on ethcore/src/trace/flat.rs:
```diff
@@ -169,15 +164,25 @@ mod tests {
+		assert_eq!(ordered_traces[0].subtraces, 2);
+		//assert_eq!(ordered_traces[1].parent, Some(0));
+		//assert_eq!(ordered_traces[1].children, vec![2, 3]);
+		assert_eq!(ordered_traces[1].trace_address, vec![0, 0]);
```
should be `vec![0]`, i think.

gavofyork commented on ethcore/src/trace/flat.rs:
```diff
@@ -169,15 +164,25 @@ mod tests {
+		assert_eq!(ordered_traces[1].subtraces, 2);
+		//assert_eq!(ordered_traces[2].parent, Some(1));
+		//assert_eq!(ordered_traces[2].children, vec![]);
+		assert_eq!(ordered_traces[2].trace_address, vec![0, 0, 0]);
```
remove first `0, ` of vec address.

gavofyork commented on ethcore/src/trace/flat.rs:
```diff
@@ -169,15 +164,25 @@ mod tests {
+		assert_eq!(ordered_traces[2].subtraces, 0);
+		//assert_eq!(ordered_traces[3].parent, Some(1));
+		//assert_eq!(ordered_traces[3].children, vec![]);
+		assert_eq!(ordered_traces[3].trace_address, vec![0, 0, 1]);
```
remove first `0, ` of vec address.

gavofyork commented on ethcore/src/trace/flat.rs:
```diff
@@ -169,15 +164,25 @@ mod tests {
+		assert_eq!(ordered_traces[3].subtraces, 0);
+		//assert_eq!(ordered_traces[4].parent, Some(0));
+		//assert_eq!(ordered_traces[4].children, vec![]);
+		assert_eq!(ordered_traces[4].trace_address, vec![0, 1]);
```
remove first `0, ` of vec address.

gavofyork commented:
first `0` of trace address looks redundant.

debris commented:
you are right! 👍
